### DISCUSSION TITLE
The -ping command line option and its effect on thumbnail creation

### DISCUSSION THREAD

**Discussion by danielwlewis** [ImageMagick/ImageMagick#7814] [Help]
### ImageMagick version

7.1.1-21

### Operating system, version and so on

Windows 11

### Description

I've written a C program that (among other things) uses Magick with a large number of JPG files to (1) create thumbnails and (2) extract EXIF metadata. I do this using scripting as follows:

First, I use the Windows CreateProcess API to load Magick into memory with piped stdin and stdout and feed it some initial options by writing to its stdin, such as:

   magick -define jpeg:size=160x200 -format %[EXIF:Make*] -script -

For each image to be processed, I begin by feeding Magick with the filename piped to its stdin. Then, if I need a thumbnail of the image, I then feed Magick something like the following to its stdin:

   -thumbnail 80x100 -write thumbnail.jpg

If I need EXIF data, I feed it something like this to its stdin (and read its stdout response from the pipe):

   -write info: 

Finally, when I'm done with this image, I feed the following to its stdin:

   -delete 0\n

My problem is this: To speed up extraction of EXIF metadata, I tried putting -ping in front of -write info:. Although it made a huge reduction in time, it caused subsequent thumbnail creations to be all black. So then I added +ping in front of -thumbnail, but still get black thumbnails.

It seems to me that if you ask Magick to create a thumbnail, it should ignore the -ping option anyway, but it doesn't. Am I doing something wrong, or is there a way to use -ping with metadata extraction without affecting thumbnail creation?

Thanks!
Dan

**Comment by snibgo** [comment]
> ... or is there a way to use -ping with metadata extraction without affecting thumbnail creation?

No. The purpose of "-ping" is to prevent reading pixels when an image is read. Without pixels, writing an image is pointless.

EDIT: next paragraph may be wrong:
Another point: "-ping" is one of the few options that has an overall effect wherever it is placed in the command. I regard this as a poor feature (I would like it to apply only to images read *after* the "-ping"), but even if that is changed, it wouldn't help you: if you have no pixels, you can't write the pixels to an image.

 > Although it made a huge reduction in time, ...

Do you mean the overall command, including reading the image, takes less time? Obviously, as no pixels are read.

Or do you mean that writing info takes less time? That would surprise me, as there are only 80x100=8000 pixels, and anyhow the processing for that format shouldn't involve the pixels. How long does it take?

**Reply by danielwlewis** [reply]
Hi snibgo,

I suspect I may not have been entirely clear. I realize that -ping avoids reading pixels. I was hoping to use it when extracting the EXIF metadata, but NOT when creating a thumbnail. E.g., there are three possibilities for any one image:

1. Only extract EXIF metadata (use -ping)
2. Only create a thumbnail (don't use -ping)
3. Do both (don't use -ping)

What I'm seeing is that once -ping is used, it continues to be in effect for subsequent images - even if not specified as an option for those subsequent images.

W.r.t. processing time, I have about 18,000 photos and almost all are about 1000x1000 pixels. (The thumbnails are 80x100 - not the original photos.) Extracting metadata (but no thumbnails) for all 18,000 takes about 8 minutes without -ping. With it, the time is cut in half.

'Hope that clarifies,
Dan

**Comment by snibgo** [comment]
Experiments suggest my response was not correct. With this:
```
magick -ping AGA_1282.JPG +ping toes.png -verbose info:
```
... the first image is reported as entirely black (because no pixels were read) but the second image has correct colours.

Note that I put -ping or +ping *before* reading the relevant image. 

> ... tried putting -ping in front of -write info:.

So that is your problem. "-ping" is a setting for *reading*, not *writing*.

**Reply by danielwlewis** [reply]
That's encouraging! 'Looks like the answer is to always place the option BEFORE the filename of the (original) photo. I'll see if I can get it working later tonight.

Thanks!
Dan

**Comment by snibgo** [comment]
Yes. For your three cases, I suggest sending this to magick:
```
-ping img1.png -write info: +delete
+ping img2.png -thumbnail 80x100 -write thumb2.jpg +delete
+ping img3.png -thumbnail 80x100 -write thumb3.jpg -write info: +delete
```

**Reply by danielwlewis** [reply]
Agreed! 'Just tested it and seems to work fine. Thanks!
Dan
